File: data-structures-algorithms/module1/week2_algorithmic_warmup/8_last_digit_of_the_sum_of_squares_of_fibonacci_numbers/fibonacci_sum_squares.py
import unittest


def fibonacci_sum_squares(n):
    """Return the last digit of the sum of the squares of the first n Fibonacci numbers"""

    # The sum of the squares of the first n Fibonacci numbers is the n-th Fibonacci number times the (n + 1)-th Fibonacci number
    # The last digit of the sum of the squares of the first n Fibonacci numbers is the last digit of the n-th Fibonacci number times the last digit of the (n + 1)-th Fibonacci number

    if n <= 1:
        return n

    previous = 0
    current = 1

    # The Pisano period for 10 is 60
    for _ in range(n % 60):
        previous, current = current, (previous + current) % 10

    return (previous * current) % 10


class TestFibonacciSumSquares(unittest.TestCase):
    """Test for fibonacci_sum_squares function"""

    def test_fibonacci_sum_squares(self):
        """Test for fibonacci_sum_squares function"""
        self.assertEqual(fibonacci_sum_squares(7), 3)
        self.assertEqual(fibonacci_sum_squares(73), 1)


if __name__ == "__main__":
    n = int(input())
    print(fibonacci_sum_squares(n))

File: data-structures-algorithms/module1/week2_algorithmic_warmup/8_last_digit_of_the_sum_of_squares_of_fibonacci_numbers/fibonacci_sum_squares.cpp
#include <iostream>

int fibonacci_sum_squares_naive(long long n) {
    if (n <= 1)
        return n;

    long long previous = 0;
    long long current  = 1;
    long long sum      = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
        sum += current * current;
    }

    return sum % 10;
}

int fibonacci_sum_squares(long long n) {
    // The sum of the squares of the first n Fibonacci numbers is the nth Fibonacci number times the (n + 1)th Fibonacci number.

    // The period of the last digit of the Fibonacci numbers is 60.
    n = n % 60;
    if (n <= 1)
        return n;

    long long previous = 0;
    long long current  = 1;

    for (long long i = 0; i < n; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = (tmp_previous + current) % 10;
    }

    return (previous * current) % 10;
}

int main() {
    long long n = 0;
    std::cin >> n;
    //std::cout << fibonacci_sum_squares_naive(n) << std::endl;
    std::cout << fibonacci_sum_squares(n) << std::endl;
}
